test_runner_lib/include/test_runner/TestOutputSpec.h

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2022 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <glog/logging.h>
+#include <remill/Arch/Arch.h>
+#include <test_runner/TestRunner.h>
+
+#include <any>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace test_runner {
+
+using MemoryModifier = std::function<void(MemoryHandler &)>;
+using RegisterValue = std::variant<uint64_t, uint8_t>;
+using RegisterValueRef = std::variant<std::reference_wrapper<uint64_t>,
+                                      std::reference_wrapper<uint8_t>>;
+
+struct RegisterCondition {
+  std::string register_name;
+  RegisterValue enforced_value;
+};
+
+template <typename T>
+concept State = std::is_base_of_v<ArchState, T>;
+
+template <State S>
+class TestOutputSpec {
+ public:
+  uint64_t addr;
+  std::string target_bytes;
+
+ private:
+  using RegisterAccessorMap =
+      std::unordered_map<std::string, std::function<RegisterValueRef(S &)>>;
+  using RegisterConditionList = std::vector<RegisterCondition>;
+  using MemoryConditionList = std::vector<MemoryModifier>;
+
+  remill::Instruction::Category expected_category;
+  RegisterConditionList register_preconditions;
+  RegisterConditionList register_postconditions;
+  MemoryConditionList initial_memory_conditions;
+  MemoryConditionList expected_memory_conditions;
+  RegisterAccessorMap reg_to_accessor;
+
+  template <typename T>
+  void ApplyCondition(S &state, const std::string reg, T value) const {
+    auto accessor = reg_to_accessor.find(reg);
+    if (accessor != reg_to_accessor.end()) {
+      std::visit([value](auto &&arg) { arg.get() = value; },
+                 accessor->second(state));
+    }
+    throw std::runtime_error(std::string("Unknown reg: ") + reg);
+  }
+
+  template <typename T>
+  void CheckCondition(S &state, const std::string reg, T value) const {
```

Comment:
Taking `const T` as argument does not make that much sense usually (such in case of `std::string`) - I am giving you a brand new copy, so you can do whatever you want with it - I guess having it as `const` helps from accidentally modifying it inside the body though - but here it looks like a `const &` would work?